iceoryx_utils/test/moduletests/test_cxx_deadline_timer.cpp

```diff
@@ -0,0 +1,133 @@
+#include "iceoryx_utils/internal/units/duration.hpp"
+#include "test.hpp"
+#include "testutils/timing_test.hpp"
+
+#include <chrono>
+#include <thread>
+
+using namespace ::testing;
+
+using namespace iox::units;
+using namespace iox::units::duration_literals;
+
+using Timer = iox::cxx::DeadlineTimer;
+
+class DeadlineTimer_test : public Test
+{
+  public:
+    virtual void SetUp()
+    {
+        numberOfCalls = 0;
+    }
+
+    virtual void TearDown()
+    {
+    }
+
+    Duration second{1_s};
+
+    std::atomic<int> numberOfCalls{0};
+    static const Duration TIMEOUT;
+    static const int SLEEPTIME;
+};
+
+const Duration DeadlineTimer_test::TIMEOUT{10_ms};
+const int DeadlineTimer_test::SLEEPTIME = DeadlineTimer_test::TIMEOUT.milliSeconds<int>();
+
+TIMING_TEST_F(DeadlineTimer_test, ZeroTimeoutTest, Repeat(5), [&] {
+    Timer sut(0_s);
+
+    TIMING_TEST_EXPECT_TRUE(sut.hasExpired());
+});
+
+TIMING_TEST_F(DeadlineTimer_test, DurationOfNonZeroIsExpiresAfterTimeout, Repeat(5), [&] {
+    Timer sut(TIMEOUT);
+
+    TIMING_TEST_EXPECT_FALSE(sut.hasExpired());
+    std::this_thread::sleep_for(std::chrono::milliseconds(2 * SLEEPTIME / 3));
+    TIMING_TEST_EXPECT_FALSE(sut.hasExpired());
+    std::this_thread::sleep_for(std::chrono::milliseconds(2 * SLEEPTIME / 3));
+    TIMING_TEST_EXPECT_TRUE(sut.hasExpired());
+});
+
+TIMING_TEST_F(DeadlineTimer_test, ResetWithDurationIsExpired, Repeat(5), [&] {
+    Timer sut(TIMEOUT);
+    std::this_thread::sleep_for(std::chrono::milliseconds(2 * SLEEPTIME));
+    TIMING_TEST_EXPECT_TRUE(sut.hasExpired());
+    sut.reset();
+    TIMING_TEST_EXPECT_FALSE(sut.hasExpired());
+});
+
+TIMING_TEST_F(DeadlineTimer_test, ResetWhenNotExpiredIsStillNotExpired, Repeat(5), [&] {
+    Timer sut(TIMEOUT);
+    std::this_thread::sleep_for(std::chrono::milliseconds(2 * SLEEPTIME / 3));
+    sut.reset();
+    std::this_thread::sleep_for(std::chrono::milliseconds(2 * SLEEPTIME / 3));
+    TIMING_TEST_EXPECT_FALSE(sut.hasExpired());
+});
+
+TIMING_TEST_F(DeadlineTimer_test, ResetAfterBeingExpiredIsNotExpired, Repeat(5), [&] {
+    Timer sut(TIMEOUT);
+    std::this_thread::sleep_for(std::chrono::milliseconds(2 * SLEEPTIME));
+
+    TIMING_TEST_ASSERT_TRUE(sut.hasExpired());
+    sut.reset();
+    TIMING_TEST_EXPECT_FALSE(sut.hasExpired());
+});
+
+TIMING_TEST_F(DeadlineTimer_test, ResetWithCustomizedTimeAfterBeingExpiredIsNotExpired, Repeat(5), [&] {
+    Timer sut(TIMEOUT);
+    std::this_thread::sleep_for(std::chrono::milliseconds(2 * SLEEPTIME));
+
+    TIMING_TEST_ASSERT_TRUE(sut.hasExpired());
+
+    sut.reset(20_s);
+    std::this_thread::sleep_for(std::chrono::milliseconds(2 * SLEEPTIME));
+
+    TIMING_TEST_EXPECT_FALSE(sut.hasExpired());
+});
+
+TIMING_TEST_F(DeadlineTimer_test, RemainingTimeCheckIfExpired, Repeat(5), [&] {
+    Timer sut(TIMEOUT);
+    std::this_thread::sleep_for(std::chrono::milliseconds(2 * SLEEPTIME));
+    int remainingTime = sut.remainingTime().milliSeconds<int>();
```

Comment:
could you please check if the timer is expired before getting the remaining time? Just to be sure if the test below fails it's not because for some weird reason the timer is not yet expired.

Reply:
Yes. Will add a check for that.

Reply:
yes added this check now